### PR TITLE
update: Enforce that Collections and Pages have jointly unique slugs

### PR DESCRIPTION
**Pull request by idreyn**
This is related to #1055 

Here we add the following:

- UI in Collection settings to change a Collection's slug
- Checks that enforce that Collections and Pages cannot share slugs (and error states for this in the UI)
- Code that guarantees that a newly created Collection or Page will get a unique slug

There are also some changes (improvements?) to error handling that we already talked about @3mcd 

_Test plan:_
Visit the settings for both a Page and a Collection and verify that you can change their slugs, but not to a slug that another object in the Community already owns. Create a new Collection and Page and verify that they both have valid slugs.

Run tests:
```
npm run test-dev server/collections/
```

Sorry for the alarming number of commits listed below...I need to get good at rebasing again with all of these stacked PRs.

### CODE REVIEW

File: client/utils/usePersistableState.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { useState } from 'react';
+import { useBeforeUnload } from 'react-use';
+
+export const usePersistableState = <T>(
```
Comment (idreyn) [inline client/utils/usePersistableState.ts:4]: This is a pattern we use everywhere, so it'll be good to have a hook for it.

Comment (3mcd) [reply client/utils/usePersistableState.ts:4]: This is really cool!

File: server/utils/errors.ts [hunk 2]
```diff
@@ -61,3 +88,18 @@ export const handleErrors = (req, res, next) => {
...
+		}
+	} else if (err instanceof HTTPStatusError) {
+		if (!res.headersSent) {
+			res.status(err.status).send(err.message);
+		}
+	} else if (!res.headersSent) {
+		res.status(500);
+		next(err);
+	}
+};
```
Comment (3mcd) [inline server/utils/errors.ts:105]: Nice!!

**Review by 3mcd — APPROVED** [review]
LGTM!! The error code looks great, too. I'll take your word for it that `findAcceptableSlug` is called in each of the sites where it's needed.

**Pull request merged**